### PR TITLE
Fixes #33492: Keep logical test suite search docs intact when recreated by upsert

### LINKED ISSUE #33492: Bundle (logical) test suites list returns 400: search doc created with `tests: {}` by updateEntity upsert
### Affected module
Backend / Search

### Describe the bug
On 2.0.x the Data Quality → Bundle Suites (logical test suites) page can fail with a 400:

```
GET /api/v1/dataQuality/testSuites/search/list?limit=15&fields=owners,summary&offset=0&includeEmptyTestSuites=true&testSuiteType=logical&sortField=lastResultTimestamp&sortType=desc

{"code":400,"message":"Cannot deserialize value of type `java.util.ArrayList<org.openmetadata.schema.type.EntityReference>` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.openmetadata.schema.tests.TestSuite[\"tests\"])"}
```

A logical suite document in the test suite search index ends up with `"tests": {}` and `"testsRevision": {}`. [`listFromSearchWithOffset`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java#L4629) converts every hit to `TestSuite`, so one bad document fails the whole page. This happens even though the call doesn't request `tests`, which [`TestSuiteRepository.clearFields`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/TestSuiteRepository.java#L308) clears afterwards.

### Root cause
Three defects combine; a fourth issue makes it easy to hit.

**1. The `updateEntity` upsert document is serialized as `{}` for every field (since 1.11.0).**
[`OpenSearchEntityManager.updateEntity`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/opensearch/OpenSearchEntityManager.java#L453-L464) and [`ElasticSearchEntityManager.updateEntity`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchEntityManager.java#L420-L431) build a `Map<String, JsonData>` and pass it to both `.params(...)` and `.upsert(...)`. The client serializes the upsert map with its plain Jackson serializer, so each `JsonData` value is written as `{}`. Serializing an `UpdateRequest` with opensearch-java 3.4.0 gives:

```json
{"script":{"params":{"name":"suite","tests":[{"id":"…","type":"testCase"}],"testsRevision":5},"source":"…"},
 "scripted_upsert":true,
 "upsert":{"name":{},"tests":{},"testsRevision":{}}}
```

The script params are correct; only the upsert document is broken. The bulk sinks already build `JsonData` correctly ([`OpenSearchBulkSink`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/searchIndex/OpenSearchBulkSink.java#L630) uses `JsonData.of(value, JACKSON_JSONP_MAPPER)` and `OsUtils.toJsonData(json)`). Before 2.0 this was masked, because every update script copied all fields from the params.

**2. The 2.0 relationship-preserving script keeps the junk values.**
For logical suites, `updateEntityIndex` [selects](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java#L1757) the [`documentUpdateScript`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java#L267-L296) and [sends it](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/SearchRepository.java#L1765) through `updateEntity`. The script skips `tests` and `testsRevision` if the document already contains either key. When the document is missing, it is created from the upsert document, which has both keys set to `{}`, so `tests: {}` and `testsRevision: {}` are kept.

**3. The index accepts the malformed values.**
- **Indices created by 2.0:** [mapping hardening](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/SearchIndexSettings.java#L31) sets `ignore_malformed: true` on numeric fields, so `testsRevision: {}` is stored in `_source`. `tests` has no explicit mapping, so `{}` is accepted as an object.
- **Indices created before 2.0:**
  - `migrate` → [`OpenSearchIndexManager.updateIndex`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/search/opensearch/OpenSearchIndexManager.java#L85) fails with `cannot change object mapping from non-nested to nested` (`owners`), and logs this without the cause. As a result `testsRevision` is never mapped.
  - The first `{}` then maps `testsRevision` dynamically as `object`. After that, every numeric `testsRevision` write to that index is rejected with `object mapping for [testsRevision] tried to parse field [testsRevision] as object, but found a concrete value`.
  - Other logical suites stop updating in search, and newly created ones never appear in the list.

**4. Search Indexing skips empty logical suites.**
[`EntityRepository.getReindexFilter()`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java#L2322) returns `new ListFilter(Include.ALL)`, and `TestSuiteRepository` does not override it. [`TestSuiteDAO.listCount/listBefore/listAfter`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java#L9328-L9395) inner-join on test case relationships unless `includeEmptyTestSuites=true`. After any Search Indexing run, every logical suite without test cases therefore has no search document, and the next edit to one of them hits defect 2. Locally, a full run for `testSuite` indexed all 17 logical suites that had test cases and none of the 4 empty ones.

Once a document is poisoned, adding test cases does not repair it: the replacement script evaluates `params.testsRevision >= ctx._source.testsRevision` against `{}` and fails with a `script_exception`.

### To Reproduce
Reproduced against a local server on the 2.0.1 code path (`main` @ 4d886c75) with OpenSearch 3.4:
1. Create a logical test suite with no test cases: `POST /api/v1/dataQuality/testSuites` with `{"name": "empty-bundle-suite"}`.
2. Run Search Indexing for test suites: `POST /api/v1/apps/trigger/SearchIndexingApplication` with `{"entities": ["testSuite"]}`, and wait for it to finish. The suite now has no document in the test suite search index.
3. Edit the suite: `PATCH /api/v1/dataQuality/testSuites/{id}` with `[{"op":"add","path":"/description","value":"x"}]`.
4. `GET <test suite index>/_doc/{id}` shows `"tests": {}` and `"testsRevision": {}`.
5. Open Bundle Suites, or call the `search/list` URL above. The page containing the suite returns the 400.
6. Optional: add a test case to the suite with `PUT /api/v1/dataQuality/testCases/logicalTestCases`. The document stays `tests: {}`, and the server logs a `script_exception`.

### Expected behavior
- Updating a logical suite whose search document is missing produces a correct document: `tests` as an array and a numeric `testsRevision`.
- Search Indexing indexes logical suites that have no test cases.
- A single malformed document does not fail the entire list page.

### Workaround
- **Recover:** run Search Indexing for Test Suite only. Every 2.0.x run rebuilds the index and swaps it in. Verified locally: all Bundle Suites pages load again.
- **Avoid recurrence until fixed:** after a Search Indexing run, add test cases to an empty bundle suite before editing its details. Verified locally: adding test cases first indexes the suite correctly, and later edits stay correct.

### Suggested fix
1. `OpenSearchEntityManager.updateEntity` / `ElasticSearchEntityManager.updateEntity`: send the upsert as a single serialized document, e.g. `JsonData.of(doc, mapper)` or `OsUtils.toJsonData(json)`, not `Map<String, JsonData>`. This is the root cause. Other entity types that go through this path with a missing document and a partial script may also get junk documents (not verified).
2. Include empty test suites in reindex: override `getReindexFilter()` in `TestSuiteRepository` to set `includeEmptyTestSuites=true`.
3. Defensive: don't read `tests` from the search `_source` in `search/list` unless it is requested (e.g. add it to [`SEARCH_FIELDS_EXCLUDE`](https://github.com/open-metadata/OpenMetadata/blob/2.0.1-release/openmetadata-service/src/main/java/org/openmetadata/service/resources/dqtests/TestSuiteResource.java#L93)).
4. `OpenSearchIndexManager.updateIndex` should log the underlying exception when the mapping update fails.

### OpenMetadata version
2.0.1. Also present on 2.0.2 and `main`. The upsert serialization defect has been present since 1.11.0; it surfaces from 2.0.0 onward.

### PR DESCRIPTION
**Pull request by ShaileshParmar11**
### Describe your changes:

Fixes #33492

The Bundle Suites page (`GET /v1/dataQuality/testSuites/search/list?testSuiteType=logical`) can fail with a 400 (`Cannot deserialize value of type ArrayList<EntityReference> from Object value … TestSuite["tests"]`) because a logical test suite's search doc ends up with `"tests": {}`. This PR fixes the two defects behind it.

**1. `updateEntity` upserts were serialized as `{}` per field**
`OpenSearchEntityManager.updateEntity` and `ElasticSearchEntityManager.updateEntity` passed the `JsonData`-wrapped script params as the upsert document, and the client writes each `JsonData` value as `{}`. When the target doc is missing, the logical suite's relationship-preserving script keeps the `tests` / `testsRevision` it finds on the upsert.
- On **OpenSearch** the doc is stored with `tests: {}`, which causes the 400 on the list.
- On **Elasticsearch** the write is rejected (`failed to parse field [testsRevision] of type [long] … '{}'`) and the update is lost.

The upsert is now built from plain values by `SearchIndexUtils.toUpsertDocument`, which drops null fields and the `fieldsToRemove` script instruction. Request building moves into a package-private `buildUpdateEntityRequest`, the same seam as the existing `buildUpdateChildrenRequest`, so it can be unit tested.

**2. Search Indexing skipped logical suites without test cases**
The `TestSuiteDAO` list/count queries inner-join test cases unless `includeEmptyTestSuites=true`, so no reindex ever indexed an empty bundle suite. Those suites disappeared from Bundle Suites, and the next edit to one of them triggered defect 1. `TestSuiteRepository.getReindexFilter()` now sets `includeEmptyTestSuites=true`. This also aligns the reader and counts with the partition cursor (`EntityDAO.getCursorAtOffset`), which never applied the join.

Docs that are already broken are repaired by running Search Indexing for Test Suite after upgrading.

#
### Type of change:
- [x] Bug fix

#
### High-level design:
Two independent, narrow fixes, each at its root cause:
- **Upsert serialization.** A shared `SearchIndexUtils.toUpsertDocument(doc)` is used by both engine managers, so the two cannot drift apart again. Script params are unchanged. The upsert keeps its previous key set (non-null fields) but now carries real values. For scripts that overwrite every field (`DEFAULT_UPDATE_SCRIPT`) the result is identical. For relationship-preserving scripts, the doc created for a missing target is now correct.
- **Reindex filter.** An override of the existing `getReindexFilter()` extension point, which is documented as "defaults to all rows". No DAO or query changes.
- **Alternatives rejected.**
  - Sending `{}` as the upsert and letting the script build the doc: this changes behaviour for partial scripts on missing docs.
  - Hardening only `search/list` against a malformed `tests`: this hides the corrupted index instead of preventing it.
- **Migration / backward compatibility.** No schema or data migration. Existing broken docs need a Search Indexing run for Test Suite.

#
#### Use cases covered
- Editing a logical test suite whose search doc is missing recreates the doc with `tests` as a list, and `search/list` keeps working.
- A Search Indexing run keeps logical test suites that have no test cases in the index.

#### Unit tests
- [x] I added unit tests for the new/changed logic.
- **Added:** `OpenSearchEntityManagerUpdateEntityTest` and `ElasticSearchEntityManagerUpdateEntityTest`. They serialize the real `UpdateRequest` with the client's `JacksonJsonpMapper`. Before the fix they failed with `upsert: {"tests":{},"name":{},"testsRevision":{}}`.
- **Re-run, all passing (254 tests):**
  - `SearchRepositoryTest`, `SearchRepositoryBehaviorTest`, `SearchIndexUtilsTest`
  - `TestSuiteIndexTest`, `TestSuiteRepositoryTest`
  - `PartitionCalculatorTest`, `DistributedIndexingStrategyTest`
  - `OpenSearchBulkSinkBehaviorTest`, `ElasticSearchBulkSinkBehaviorTest`
  - `*EntityManagerUpdateChildrenTest`, `UpdateChildrenAsyncPropagationTest`
- **Coverage:**
  - The new unit tests cover the new `buildUpdateEntityRequest` in both managers and `SearchIndexUtils.toUpsertDocument`.
  - `TestSuiteReindexIT` covers `TestSuiteRepository.getReindexFilter`.
  - Whole-class JaCoCo % was not measured: the two entity managers are large and exercised mainly by ITs.

#### Backend integration tests
- [x] I added integration tests in `openmetadata-integration-tests/`.
- `TestSuiteResourceIT#test_updateLogicalSuiteWithMissingSearchDocKeepsTestsList`: deletes a logical suite's search doc, patches the suite, then asserts `tests` is indexed as a list and `search/list` returns the suite.
- `TestSuiteReindexIT#reindexKeepsLogicalTestSuiteWithoutTestCases` (new, `@Isolated`, global-state lane): creates an empty logical suite, runs a `testSuite`-only Search Indexing run, then asserts the suite is in the promoted index.

Run locally with and without the fix:

| IT | mysql-elasticsearch (with fix / without) | postgres-opensearch (with fix / without) |
|---|---|---|
| missing search doc + edit | pass / pass* | pass (CI) / **fail** (`tests must be indexed as a list, was: {}`) |
| reindex keeps empty suite | pass / **fail** (`expected: <1> but was: <0>`) | CI (global-state lane)** / not run |

\* On Elasticsearch the pre-fix upsert is rejected rather than stored, and the retry queue later rebuilds the doc, so this IT only tells the two cases apart on OpenSearch (the `postgres-opensearch` CI lane).

\*\* The local run with the fix on `postgres-opensearch` was invalidated by an OOM-killed test container, so CI covers it. `TestSuiteReindexIT` runs in the isolated (global-state) lane next to `AppsResourceIT`. In the parallel lane, K3s is started on demand for pipeline tests, and the test's K8s assumption would skip it.

#### Ingestion integration tests
- [x] Not applicable (no ingestion changes).

#### Playwright (UI) tests
- [x] Not applicable (no UI changes).

#### Manual testing performed
1. **Reproduced end to end before the fix** on a local 2.0.x stack (OpenSearch 3.4):
   1. Created a logical test suite with no test cases.
   2. Ran Search Indexing for `testSuite`; the suite had no search doc afterwards.
   3. Patched its description; the doc was stored with `"tests": {}` and `"testsRevision": {}`.
   4. `testSuites/search/list?testSuiteType=logical&fields=owners,summary` returned the reported 400.
   5. Adding a test case afterwards did not repair the doc.
2. **Checked the recovery path:** a Search Indexing run for `testSuite` repairs an already-broken index, and every list page returns 200.
3. **Validated the fixed server** through the integration tests above (locally on `mysql-elasticsearch`; `postgres-opensearch` via CI). It was not deployed to the local Docker stack.

#
### UI screen recording / screenshots:

Not applicable.

#
### Checklist:
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] My PR title is `Fixes <issue-number>: <short explanation>`
- [x] My PR is linked to a GitHub issue via `Fixes #<issue-number>` above.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] For JSON Schema changes: not applicable, no schema changes.
- [x] For UI changes: not applicable.
- [x] I have added tests (unit / integration / Playwright as applicable) and listed them above.
- [x] I have added a test that covers the exact scenario we are fixing. For complex issues, comment the issue number in the test for future reference.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h2><a href="https://app.greptile.com/api/retrigger?id=65479915"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/RetriggerDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/Retrigger.svg?v=2">Retrigger</picture></a>Confidence Score: 4/5</h2>

The functional fix appears sound, but the explicit integration-test execution-mode requirement must be satisfied before merging.

<h2>Findings</h2>

1. P2&nbsp;**Wrong Integration Test Mode** <a href="https://github.com/open-metadata/OpenMetadata/pull/33498#discussion_r4037666850">▶</a>

<details open><summary>Summary</summary>

This PR prevents malformed or missing logical test-suite search documents by correcting scripted-upsert serialization and including empty suites during reindexing.
- Builds Elasticsearch and OpenSearch upserts from plain, non-null document values while excluding script-only removal instructions.
- Configures test-suite reindexing to include logical suites without test cases.
- Adds unit and integration coverage for missing-document updates and empty-suite reindexing.
</details>

<details open><summary>Diagram</summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
  DB[(Metadata database)] --> RF[Test-suite reindex filter]
  RF -->|include empty suites| IDX[(Search index)]
  PATCH[Test-suite update] --> DOC[Plain upsert document]
  DOC -->|missing target document| IDX
  IDX --> LIST[Logical suite search/list]
```
</details>

<sub>Reviews (5) · Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/open-metadata/openmetadata/commit/602e2b7920d7a6327380299b99d0b055de50ca7e)</sub>

### CODE REVIEW

File: openmetadata-integration-tests/src/test/java/org/openmetadata/it/tests/TestSuiteReindexIT.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    assumeFalse(
+        TestSuiteBootstrap.isK8sEnabled(), "App trigger not compatible with K8s pipeline backend");
+    TestSuite emptySuite =
+        SdkClients.adminClient()
+            .testSuites()
+            .create(new CreateTestSuite().withName(ns.prefix("empty_logical_suite")));
+
+    reindexTestSuites();
+
+    try (Rest5Client searchClient = TestSuiteBootstrap.createSearchClient()) {
```
[the remark this answers is not in the document]
Comment (ShaileshParmar11) [reply openmetadata-integration-tests/src/test/java/org/openmetadata/it/tests/TestSuiteReindexIT.java:68]: Added the `OssTestServer.isExternalMode()` assumption to `TestSuiteReindexIT` (4442c05), matching `EntityShapeIT`; it is the one that goes through `OssTestServer`. Left `TestSuiteResourceIT` unchanged: that class already calls `TestSuiteBootstrap.createSearchClient()` unguarded in its existing search-index tests (43 such calls across `tests/`), and that lane runs embedded-only.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: openmetadata-integration-tests/src/test/java/org/openmetadata/it/tests/TestSuiteReindexIT.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import org.openmetadata.schema.utils.JsonUtils;
+import org.openmetadata.service.Entity;
+
+/**
+ * A logical test suite without test cases must survive a Search Indexing run. The test suite DAO
+ * drops empty suites unless {@code includeEmptyTestSuites=true}, so the reindex left them out of
+ * the rebuilt index; the next update to such a suite recreated its doc from an upsert (#33492).
+ */
+@Execution(ExecutionMode.SAME_THREAD)
+@Isolated
```
[the remark this answers is not in the document]
Comment (ShaileshParmar11) [reply openmetadata-integration-tests/src/test/java/org/openmetadata/it/tests/TestSuiteReindexIT.java:40]: Keeping `SAME_THREAD`. `TestSuiteReindexIT` triggers the server-wide Search Indexing app and runs in the global-state lane, where the other app-triggering ITs (`AppsResourceIT`, `SystemResourceIT`, `WorkflowDefinitionResourceIT`) and `SearchIndexPromotionIT` all use `@Execution(ExecutionMode.SAME_THREAD)` with `@Isolated`. The `CONCURRENT` convention applies to the parallel-lane entity ITs.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: openmetadata-service/src/main/java/org/openmetadata/service/search/SearchIndexUtils.java [hunk 3]
```diff
@@ -139,6 +140,19 @@ public static String stripLineageForSize(
...
 
+  /**
+   * The document a scripted update stores when its target doc is missing. It must hold plain values:
+   * the {@code JsonData}-wrapped script params serialize as {@code {}} per field when used as the
+   * upsert, and scripts that preserve existing fields (e.g. a logical test suite's {@code tests})
+   * then keep those empty objects (#33492).
+   */
+  public static Map<String, Object> toUpsertDocument(Map<String, Object> doc) {
+    Map<String, Object> upsert = JsonUtils.getMap(doc);
+    upsert.values().removeIf(Objects::isNull);
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/SearchIndexUtils.java:151]: We need to double check if removing a null value from the doc can create undesired effects where in AI Analytics or in other place a we would be looking for explicit null (vs an absent key)

Comment (ShaileshParmar11) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/SearchIndexUtils.java:151]: Good call to check. I verified this on OpenSearch 3.4 with a probe that sends the exact pre-fix and post-fix request bodies to a missing doc. The bodies come from the real `convertToJsonDataMap` / `buildUpdateEntityRequest`, run with the real `DEFAULT_UPDATE_SCRIPT` and logical-suite script. The input doc has `description: null`, `tier: null` and a null nested in `tests[0]`.

| Script | Stored before the fix | Stored after the fix |
|---|---|---|
| `DEFAULT_UPDATE_SCRIPT` | `{basic, name, tests: [...], testsRevision: 5, fieldsToRemove: {}}` | `{basic, name, tests: [...], testsRevision: 5}` |
| logical-suite script | `{basic, name, tests: {}, testsRevision: {}, fieldsToRemove: {}}` | `{basic, name, tests: [...], testsRevision: 5}` |

Neither version stores explicit nulls. Null fields are absent both before and after: `convertToJsonDataMap` already dropped them from the old upsert, and `JsonUtils.getMap` drops nested nulls. Script params are identical before and after.

The only differences are the fix itself (real values instead of `{}`) and that docs created this way no longer get a junk `fieldsToRemove: {}` field, which is a script instruction, not document data. Keeping nulls in the upsert would be the behavior change: it would start creating explicit-null fields that never existed.

**Review by TeddyCr — APPROVED** [review]
